### ISSUE TITLE
🐛 BUG: D1 migrations/executions with multiple statements & quote marks in comments fail obscurely on remote databases

### ISSUE THREAD

**Issue by huw** [cloudflare/workers-sdk#4713]
### Which Cloudflare product(s) does this pertain to?

D1

### What version(s) of the tool(s) are you using?

3.22.3 [Wrangler]

### What version of Node are you using?

20.9.0

### What operating system are you using?

Linux (Debian 12 Bookworm)

### Describe the Bug

Create the following SQL file:

```sql
-- It's dare
select 1;
select 1;
```

Call `wrangler d1 execute database --file ./test.sql`

You’ll see:

```
🌀 Mapping SQL input into an array of statements
🌀 Parsing 2 statements
🌀 Executing on remote database database (<account_id>):
🌀 To execute on your local development database, pass the --local flag to 'wrangler d1 execute'

✘ [ERROR] A request to the Cloudflare API (/accounts/<account_id>/d1/database/<database_id>/query) failed.

  A prepared SQL statement must contain only one statement. [code: 7500]
  
  If you think this is a bug, please open an issue at:
  https://github.com/cloudflare/workers-sdk/issues/new/choose
```

Execution succeeds if you execute it on `--local`, _or_ if there’s only one statement. This is likely a bug in the multi-statement parser. It’s also just hard to debug, so I wanted to make this issue for people to land at if they experience it in the future.

**Comment by serinuntius** [comment]
same issue.

I removed a multi-line comment (`/*    ~~~~~~~~   */`) and it worked.

**Comment by AverageHelper** [comment]
Also facing this issue. My SQL executes in local but not remote. Literally all I did to fix was remove a `'s` ending to a word in a `--` comment, and now the whole file executes correctly in both environments.

**Comment by andrew-pyle** [comment]
I've encountered error code 7500 as well. Error was resolved by removing all SQL comments except the first automatically generated comment line.

```sql
-- Migration number: 0001 	 2024-03-28T21:27:10.888Z ← ✅ OK
-- ← ❌ FAILS
-- comment here ← ❌ FAILS
```

```sh
$ npx wrangler d1 migrations apply <database> --remote
✘ [ERROR] A prepared SQL statement must contain only one statement. [code: 7500]
```

It is possible that this is related to D1's serialization method. If newlines are removed naively,  part of the valid SQL code could be sucked back onto another line which begins with a `-- `, like a SQL injection on yourself.

I got this idea because it happened to me when I copied and pasted SQL with comments into the D1 dashboard's console. It transformed the whole SQL command into a single line, and only the first line or two of SQL survived before the rest was commented out.

**Comment by Veloxious** [comment]
> Also facing this issue. My SQL executes in local but not remote. Literally all I did to fix was remove a `'s` ending to a word in a `--` comment, and now the whole file executes correctly in both environments.

Literally all I had to do to fix the same issue. Worked locally but not in remote. Simply removing the `'s` I had in an inline comment resolved my error. I do not however have any multiline comments.

**Comment by jackrdye** [comment]
Issue still present, comments in .sql file cause execution to fail. sql parser should be fixed for d1. Plenty of people will require sql scripts.

**Comment by broofa** [comment]
I appear to be running into this same issue using the `sql` query via the [`query` REST endpoint](https://developers.cloudflare.com/api/operations/cloudflare-d1-query-database).

This SQL works:
```sql
CREATE TABLE IF NOT EXISTS users (userID INTEGER);
```
... and this SQL works:
```sql
CREATE TABLE IF NOT EXISTS sessions (sessionID TEXT);
```
... but this SQL does not: (fails with `400 Bad Request`:
```sql
CREATE TABLE IF NOT EXISTS users (userID INTEGER);
CREATE TABLE IF NOT EXISTS sessions (sessionID TEXT);
```
Similarly...
```sql
PRAGMA table_list; PRAGMA table_list;

```
... also fails.

tl;dr: From where I sit, multi-statement SQL queries are completely broken.

**Comment by edmundhung** [comment]
Closing this as fixed for `wrangler d1 execute --file`.

Remote file execution now uses the D1 import API instead of the old split-and-send-statements path, and Wrangler’s SQL splitter now ignores inline/block comments when splitting SQL statements.

If this still happens with `wrangler d1 migrations apply --remote` on current Wrangler, please open a new issue with an updated migration file repro.
